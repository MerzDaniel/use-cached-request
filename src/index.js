
export {useCachedRequest, RequestProvider} from './use-cached-request'
export {useRequest} from './use-request'