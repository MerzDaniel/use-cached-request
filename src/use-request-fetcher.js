import React from "react";

const pendingCache = {}

export function useRequestFetcher(url, requests, setRequests, httpClient) {
  React.useEffect(() => {
    if (
      pendingCache[url]
      || (requests.data[url] && !(requests.states[url] && requests.states[url].refetch))
    ) {
      // already pending or loaded
      return
    }

    pendingCache[url] = true
    setRequests({
      data: requests.data,
      states: {
        ...requests.states,
        [url]: {
          pending: true,
          refetch: false,
        },
      },
    })

    httpClient.get(url).then(
      ({data: responseData}) => {
        pendingCache[url] = false
        setRequests({
          data: {
            ...requests.data,
            [url]: responseData,
          },
          states: {
            ...requests.states,
            [url]: {pending: false},
          },
        })
      },
    ).catch((e) => {
      pendingCache[url] = false
      setRequests({
        data: {...requests.data},
        states: {
          ...requests.states,
          [url]: {pending: false, error: e},
        },
      })
    })
  }, [url, requests.states[url]])

  function refetch() {
    setRequests({
      data: requests.data,
      states: {
        ...requests.states,
        [url]: {...requests.states[url], refetch: true},
      },
    })
  }

  return {
    error: requests.states[url] ? requests.states[url].error : null,
    isPending: requests.states[url] ? requests.states[url].pending : true,
    data: requests.data[url],
    refetch,
  }
}