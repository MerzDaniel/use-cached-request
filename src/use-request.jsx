import React from 'react'
import axios from 'axios'

const httpClient = axios
const pendingCache = {}

const useRequest = (url) => {
  const [requests, setRequestStates] = React.useState({
    states: {},
    data: {},
  })

  React.useEffect(() => {
    // already pending or loaded
    if (
      pendingCache[url]
      || (requests.data[url] && !(requests.states[url] && requests.states[url].refetch))
    ) {
      return
    }

    pendingCache[url] = true
    setRequestStates({
      data: requests.data,
      states: {
        ...requests.states,
        [url]: {
          pending: true,
          refetch: false,
        },
      },
    })

    httpClient.get(url).then(
      ({data: responseData}) => {
        pendingCache[url] = false
        setRequestStates({
          data: {
            ...requests.data,
            [url]: responseData,
          },
          states: {
            ...requests.states,
            [url]: {pending: false},
          },
        })
      },
    ).catch((e) => {
      pendingCache[url] = false
      setRequestStates({
        data: {...requests.data},
        states: {
          ...requests.states,
          [url]: {pending: false, error: e},
        },
      })
    })
  }, [url, requests.states[url]])

  function refetch() {
    setRequestStates({
      data: requests.data,
      states: {
        ...requests.states,
        [url]: {...requests.states[url], refetch: true},
      },
    })
  }

  return {
    error: requests.states[url] ? requests.states[url].error : null,
    isPending: requests.states[url] ? requests.states[url].pending : true,
    data: requests.data[url],
    refetch,
  }
}

export {
  useRequest,
}
