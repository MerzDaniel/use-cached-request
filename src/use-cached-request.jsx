import React from 'react'
import axios from 'axios'
import * as PropTypes from 'prop-types'
import {useRequestFetcher} from "./use-request-fetcher";

const httpClient = axios
const defaultState = {
  states: {},
  data: {},
}
const requestContext = React.createContext()
const RequestProvider = ({children}) => {
  const [state, setState] = React.useState(defaultState)

  return (
    <requestContext.Provider value={[state, setState]}>
      {children}
    </requestContext.Provider>
  )
}
RequestProvider.propTypes = {
  children: PropTypes.node.isRequired,
}

function useCachedRequest(url) {
  const [requests, setRequestStates] = React.useContext(requestContext) || [undefined, undefined]

  const {
    refetch, isPending, error, data
  } = useRequestFetcher(url, requests, setRequestStates, httpClient)

  return {
    refetch, isPending, error, data
  }
}

export {
  RequestProvider,
  useCachedRequest,
}
