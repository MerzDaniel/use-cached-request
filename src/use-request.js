import React from 'react'
import axios from 'axios'
import {useRequestFetcher} from "./use-request-fetcher";

const httpClient = axios

const useRequest = (url) => {
  const [requests, setRequestStates] = React.useState({
    states: {},
    data: {},
  })

  useRequestFetcher(url, requests, setRequestStates, httpClient)

  function refetch() {
    setRequestStates({
      data: requests.data,
      states: {
        ...requests.states,
        [url]: {...requests.states[url], refetch: true},
      },
    })
  }

  return {
    error: requests.states[url] ? requests.states[url].error : null,
    isPending: requests.states[url] ? requests.states[url].pending : true,
    data: requests.data[url],
    refetch,
  }
}

export {
  useRequest,
}
