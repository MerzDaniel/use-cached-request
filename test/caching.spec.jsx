import {RequestProvider, useCachedRequest} from "../src/use-cached-request";
import React from "react";
import sinon from "sinon";
import axios from "axios";
import {act, create} from "react-test-renderer";
import {waitForIO} from "./utils/wait-for-io";
import './utils/ignore-comp-update-outside-act-warning'

function TestComp({urls = ['/api', '/api']}) {
  function TestCompWithHook({url}) {
    const request = useCachedRequest(url)
    const {data, isPending, error} = request
    const res = `${isPending || data || error}`
    return <div data-request={request}>
      {res}
    </div>
  }

  return <RequestProvider>
    {
      (urls).map((u, i) => (<TestCompWithHook key={i} url={u}/>))
    }
  </RequestProvider>
}

describe('Caching', () => {
  let sandbox
  beforeEach(() => {
    sandbox = sinon.createSandbox()
  })
  afterEach(() => {
    sandbox.restore()
  })
  it('should fetch data only once', async () => {
    let stub = sandbox.stub(axios, 'get').resolves({data: 'foo-bar-baz'})
    let comp
    act(() => {
      comp = create(<TestComp/>)
    })
    await waitForIO()
    expect(stub.callCount).toBe(1)
    expect(comp.toJSON()).toMatchSnapshot()
  })
})
