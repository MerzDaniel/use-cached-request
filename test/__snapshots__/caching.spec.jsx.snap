// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Caching should fetch data only once 1`] = `
Array [
  <div
    data-request={
      Object {
        "data": "foo-bar-baz",
        "error": undefined,
        "isPending": false,
        "refetch": [Function],
      }
    }
  >
    foo-bar-baz
  </div>,
  <div
    data-request={
      Object {
        "data": "foo-bar-baz",
        "error": undefined,
        "isPending": false,
        "refetch": [Function],
      }
    }
  >
    foo-bar-baz
  </div>,
]
`;
